frontends/p4/sideEffects.h

```diff
@@ -83,6 +83,32 @@ class SideEffects : public Inspector {
         expression->apply(se);
         return se.nodeWithSideEffect != nullptr;
     }
+    /// @return true if the expression may have side-effects.
+    static bool hasSideEffect(const IR::Expression* exp,
+                      ReferenceMap* refMap,
+                      TypeMap* typeMap) {
+        auto mce = exp->to<IR::MethodCallExpression>();
+        if (mce == nullptr)
+            return false;
+        // mce does not produce a side effect in the two cases:
+        //  * isValid()
+        //  * function with all in parameters
+        auto mi = MethodInstance::resolve(mce, refMap, typeMap);
+        if (mi->is<FunctionCall>()) {
```

Comment:
This is fine, but you can also add to this list extern methods and extern functions that have the right annotations. We can merge as it is, or we can do this in a separate pull request.